### ISSUE TITLE
Master GUI; Terminal window

### ISSUE THREAD

**Issue by MagnaCapax** [MagnaCapax/PMSS#28]
We should add a terminal window into the master gui to quickly access shell even from browser.

There's plenty of different open source terminal applications, need to pick the best one first.

Not sure if this should be main feature (on spotlight) or a little bit off the beaten path, not every user needs terminal.

**Comment by MagnaCapax** [comment]
## Adversarial Review — Status Update

**Issue:** #28 — Master GUI; Terminal window
**Updated:** 2026-04-09
**Prior review:** 2026-04-02 (see comment above)
**Verdict (unchanged):** BLOCKED — superseded by #326; implementation pending

---

### Status at 7-Day Checkpoint

The blocking dependency (#326) has seen activity since the prior review, but the feature remains unimplemented.

**What changed in #326:**

- Commit `157aacb` merged ADR 0015 — browser console defer criteria are now documented. This resolves the "ADR required before code ships" blocker identified in the prior analysis. One of five pre-ship blockers is closed.
- Core feature is **not present in PMSS git/main** as of 2026-04-08 (confirmed by QA sweep): no ttyd integration, no web panel UI button, no per-user security isolation. The `complete-verify` label currently on #326 reflects ADR progress only, not feature completion.

**Remaining blockers (four of five outstanding):**

1. ~~ADR documentation~~ — ✓ ADR 0015 merged
2. `scripts/lib/update/apps/ttyd.php` binary installer (download, SHA256 verify, deploy) — pending
3. `/run/ttyd/` directory lifecycle: tmpfs recreation on boot; `chmod 1777` sticky-bit for per-account socket isolation — pending
4. Stale PID/socket handling on reboot (treat missing socket + present PID as clean-start, not error) — pending
5. `console.php` spawn helper with mandatory `escapeshellarg()` on all user-derived arguments and username re-validation at the spawn boundary — pending

**Conclusion:** No change to the BLOCKED verdict. When #326's core feature ships and QA verification passes all acceptance criteria, close #28 as a duplicate. Until then, #326 is the canonical tracking issue.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Adversarial Review — Status Update

**Issue:** #28 — Master GUI; Terminal window
**Updated:** 2026-04-17
**Prior reviews:** 2026-04-02 (initial), 2026-04-09 (7-day checkpoint)
**Verdict (unchanged):** BLOCKED — superseded by #326; ADR 0015 formally defers the feature

---

### Status at 8-Day Checkpoint

No implementation has landed. Confirmed via:
- `grep -rn "ttyd|gotty|html5.*shell|websocket.*terminal" scripts/ etc/` — zero results (QA run, 2026-04-14)
- Git log since 2026-04-09 — zero #326-tagged commits beyond the existing ADR commit (`157aacbd`)

**ADR 0015 now formally codifies the defer criteria.** The decision record (`docs/adr/0015-defer-browser-console-until-packaged-and-reviewed.md`) defines five mandatory prerequisites before any browser shell code ships:

1. A specific terminal gateway selected and approved for supported Debian baselines — *not met*
2. Gateway provisioned through a PMSS-managed path, not an ad-hoc runtime download — *not met*
3. Proxy contract lives in per-user lighttpd `custom.d/`, per ADR 0009 — *not met (no code exists)*
4. Security review covering auth reuse, loopback binding, logging, timeout, and blast radius — *not met*
5. Hermetic development tests covering status detection, disabled-path behavior, and generated proxy/process artifacts — *not met*

Zero of five prerequisites are satisfied. The `complete-verify` label currently on #326 reflects the ADR commit only, not the feature implementation. The 2026-04-14 QA run confirmed this gap explicitly.

**One concern flagged for the maintainer:** The 2026-04-16 sweep on #326 posted "🔓 UNLOCKED — Verification complete. See sweep summary." but no detailed verification result followed. Given three prior sweep runs crashed or produced thin output (2026-04-03, 2026-04-05, 2026-04-09, 2026-04-15 skipped), it is possible the 2026-04-16 run also did not verify the implementation layer. The `complete-verify` label should not be trusted as implementation confirmation until a QA run explicitly tests the code paths defined in ADR 0015 §5.

---

### Remaining blockers before #28 can close (same four as April 9)

1. `scripts/lib/update/apps/ttyd.php` — binary installer (download, SHA256 verify, deploy to `/usr/local/bin/ttyd`)
2. `/run/ttyd/` directory lifecycle — tmpfs recreation on boot; `chmod 1777` sticky-bit for per-account socket isolation
3. Stale PID/socket handling on reboot — treat missing socket + live PID file as clean-start, not error
4. `console.php` spawn helper — `escapeshellarg()` on all user-derived args, `pmssUsernameIsValid()` re-validation at spawn boundary (ADR 0004 + 0005 compliance)

---

### Verdict

BLOCKED. No change. When #326 passes QA against the ADR 0015 acceptance criteria, close #28 as a duplicate. Until then, #326 is the canonical tracking issue and the ADR is the authoritative specification for what "done" means.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Adversarial Review — Status Update

**Issue:** #28 — Master GUI; Terminal window
**Updated:** 2026-04-24
**Prior reviews:** 2026-04-02 (initial), 2026-04-09 (7-day checkpoint), 2026-04-17 (8-day checkpoint)
**Verdict (unchanged):** BLOCKED — superseded by #326; ADR 0015 formally defers the feature

---

### Status at 7-Day Checkpoint

The April 17 review flagged a concern: the `complete-verify` label on #326 appeared to reflect ADR progress only, not feature implementation. That concern is now confirmed.

**QA run on 2026-04-19 returned FAIL on #326.** The result:
- `docs/adr/0015-defer-browser-console-until-packaged-and-reviewed.md` is the only artefact on the `157aacbd` commit tagged for #326
- The ADR formally records **deferral** (Option C: defer until packaged and reviewed) — it explicitly documents non-implementation as the decision
- Zero lines of feature code were added

**Git log as of 2026-04-24**: no ttyd, no web terminal, no `console.php`, no `scripts/lib/update/apps/ttyd.php` in the repository. The three commits since April 17 are unrelated fixes.

**Blockers: 0 of 5 prerequisites met** (same as April 17):
1. Specific terminal gateway selected and approved for supported Debian baselines — *pending*
2. Gateway provisioned through a PMSS-supported path — *pending*
3. Proxy contract in per-user lighttpd `custom.d/` (ADR 0009) — *pending*
4. Security review: auth reuse, loopback binding, logging, timeout, blast radius — *pending*
5. Hermetic tests covering status detection, disabled-path, proxy/process artifacts — *pending*

**Conclusion:** No change to BLOCKED verdict. ADR 0015 is the authoritative "done" definition for #326 — when all five prerequisites are met and the QA run against the ADR 0015 acceptance criteria passes, close #28 as a duplicate of #326.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Adversarial Review — Status Update

**Issue:** #28 — Master GUI; Terminal window
**Updated:** 2026-05-01
**Prior reviews:** 2026-04-02 (initial), 2026-04-09, 2026-04-17, 2026-04-24
**Verdict (unchanged):** BLOCKED — superseded by #326; ADR 0015 formally defers the feature

---

### Status at 7-Day Checkpoint

No implementation has landed. Confirmed via:
- `grep -rn "ttyd|gotty|html5.*shell|websocket.*terminal|console\.php" scripts/ etc/` — zero results (run 2026-05-01)
- Git log since 2026-04-24: four new commits (`a32618d1`, `93d403d8`, `a4337bf6`, `8e766a9f`) — none tagged for #326 or related to browser terminal

**Blockers: 0 of 5 prerequisites met** (unchanged since April 17):
1. Specific terminal gateway selected and approved for supported Debian baselines — *pending*
2. Gateway provisioned through a PMSS-supported path — *pending*
3. Proxy contract in per-user lighttpd `custom.d/` (ADR 0009) — *pending*
4. Security review: auth reuse, loopback binding, logging, timeout, blast radius — *pending*
5. Hermetic tests covering status detection, disabled-path, proxy/process artifacts — *pending*

---

### New Finding: Close This Issue Now

This is the fifth consecutive adversarial review of #28 with an identical BLOCKED verdict and no new information. That pattern is itself a signal.

**#28 is already labeled `duplicate`.** The canonical tracking issue is #326. Having both open produces nothing — every re-review of #28 confirms it is blocked on #326, adds a comment, and exits. That loop has no end condition until #326 either ships or is formally closed.

The correct action is to **close #28 as a duplicate of #326 now**, not when #326 ships. Closing a superseded issue does not cancel the feature — it reduces tracking noise and ends the re-review cycle. When #326's prerequisites are met and the feature ships, it closes on its own merits.

**Recommendation to maintainer:** Close #28 as a duplicate of #326. #326 carries the full specification, the ADR 0015 acceptance criteria, and the tracking label. #28 adds nothing that #326 does not already cover.

---

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Closing as a duplicate of #326

This issue asks for a browser-based terminal in the web panel. That feature is fully specified and actively tracked in **#326 — "webpanel: HTML5 shell console access for users (no SSH required)"**, which is open and carries the complete use case, architecture, security model, and the terminal-tool selection. The open question in this issue's original wording — *which* tool, and how prominently to surface it — is resolved within #326's scope.

The direction is set by **ADR-0015 ("Defer browser shell console until a terminal gateway is packaged and reviewed")**, which is accepted: the feature is intentionally deferred until a vetted terminal gateway is packaged for the supported Debian baselines, provisioned through a managed update path, proxied inside per-user lighttpd (per ADR-0009), security-reviewed, and covered by hermetic tests. Those prerequisites are not yet met, so no terminal code has shipped — by design.

Prior reviews on this thread already converged on the same conclusion: this issue is a less-specified duplicate of #326 and adds nothing #326 does not already cover. **Closing this one consolidates tracking onto #326 — the feature itself remains tracked there and is unaffected by this close.**

Closing as a duplicate of #326.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Investigation Report

**Issue:** #28 — Master GUI; Terminal window
**Investigated:** 2026-06-03
**Status:** PARTIAL — the feature is fully specified and tracked under #326; implementation is intentionally deferred by ADR-0015
**Recommendation:** Close as a duplicate of #326

### Summary

#28 (a browser terminal in the web panel) is the original, less-specified version of **#326 — "webpanel: HTML5 shell console access for users (no SSH required)."** #326 is open and carries the complete specification: terminal-tool selection, security model, and scope. It is governed by **ADR-0015**, which is *Accepted* and intentionally defers implementation until five prerequisites are met. Five prior reviews on this thread converged on "close as a duplicate of #326," and a 2026-06-02 comment announced the closure — but the close was never actually applied; the issue stayed open. This pass completes it.

### What the issue asks for

A browser-accessible shell/terminal in the user web panel, plus a decision on which terminal tool to use and how prominently to surface it. Both questions are resolved inside #326's scope (ttyd selected; surfaced as an optional, opt-in panel entry rather than a spotlight feature).

### Implementation status

- **#326** — open; the canonical, fully-specified version of this request.
- **ADR-0015** (`docs/adr/0015-defer-browser-console-until-packaged-and-reviewed.md`) — *Accepted*; adopts "defer until packaged and reviewed." Prerequisites: an approved terminal gateway for the supported Debian baselines, provisioning through a managed update path (not an ad-hoc runtime download), the proxy contract in per-user lighttpd per ADR-0009, a security review (auth reuse, loopback binding, logging, timeout, blast radius), and hermetic tests.
- **Code** — not present in `main`. No `etc/skel/www/console.php` and no `scripts/lib/update/apps/ttyd.php`; verified again this pass.

So the feature is *designed and deferred*, not *built* — and that is by design, per ADR-0015, not neglect.

### Cross-issue analysis

- **#326** — canonical successor; #28 is its duplicate. Open and still maintained (last review 2026-05-07).
- **#15** (Master GUI refactoring) and **#57** (remote desktop container) sit in the same panel/remote-access area but are distinct scopes, not duplicates of #28.

### Devil's advocate

Could #28 stay open as the "user-facing tracking issue" while #326 holds the spec? No — that splits one feature across two trackers and is exactly what produced six repeated re-reviews here. One canonical issue (#326), governed by one ADR (0015), is the simpler and correct structure. The `duplicate` label on #28 already reflects this.

### Recommendation

**Close #28 as a duplicate of #326.** No separate implementation should happen here: #326 is the single tracking issue, ADR-0015 is the authoritative definition of "done," and the feature ships there when its prerequisites are satisfied. Closing #28 consolidates tracking and ends the repeated re-review cycle. The feature itself is unaffected by this close and remains tracked in #326.

Closing as a duplicate of #326.

— Väinämöinen
